parsers/TR.py

```diff
@@ -1,177 +1,162 @@
+    if r.status_code == 200:
+        return r.json()["body"][KINDS_MAPPING[kind]["json_key"]]
+    else:
+        raise ParserException(
+            parser="TR.py",
+            message=f"{target_datetime}: {kind} data is not available for TR",
+        )
 
 
-def get_last_data_idx(productions) -> int:
-    """
-    Find index of the last production.
-    :return: (int) index of the newest data or -1 if no data (empty day).
-    """
-    for i in range(len(productions)):
-        if productions[i]["total"] < 1000:
-            return i - 1
-    return len(productions) - 1  # full day
+def validate_production_data(
+    data: list,
+    logger: Logger = getLogger(__name__),
+) -> list:
+    """detects outliers: for real-time data the latest data point can be completely out of the expected range and needs to be excluded"""
+    required = [mode for mode in PRODUCTION_MAPPING]
+    expected_range = {
+        "coal": (1000, 60000),
+        "gas": (1000, 60000),
+    }
+    all_data_points_validated = [
+        x
+        for x in data
+        if validate(x, logger, required=required, expected_range=expected_range)
+    ]
+    return all_data_points_validated
 
 
-def fetch_price(
+@refetch_frequency(timedelta(days=1))
+def fetch_production(
     zone_key: str = "TR",
-    session: Optional[Session] = None,
+    session: Session = Session(),
     target_datetime: Optional[datetime] = None,
     logger: Logger = getLogger(__name__),
-):
-    if target_datetime:
-        raise NotImplementedError("This parser is not yet able to parse past dates")
-
-    zonekey.assert_zone_key(zone_key, "TR")
-
-    r = session or Session()
-    soup = BeautifulSoup(r.get(PRICE_URL).text, "html.parser")
-    cells = soup.select(".TexAlCenter")
-
-    # data is in td elements with class "TexAlCenter" and role "gridcell"
-    data = list()
-    for cell in cells:
-        if cell.attrs.get("role", "") != "gridcell":
-            continue
-        data.append(cell.text)
-
-    dates = [
-        datetime.strptime(val, "%d/%m/%Y").date()
-        for i, val in enumerate(data)
-        if i % 5 == 0
-    ]
-    times = [
-        datetime.strptime(val, "%H:%M").time()
-        for i, val in enumerate(data)
-        if i % 5 == 1
-    ]
-    prices = [val.replace(",", ".") for i, val in enumerate(data) if i % 5 == 2]
-    # sometimes the price will now have 2 decimals in it, e.g. 4.299.99
-    # therefore below removes decimals which are not the last one, e.g. changing to 4299.99 to avoid conversion error
-    prices = [float(val.replace(".", "", (val.count(".") - 1))) for val in prices]
-
-    datapoints = [
-        {
-            "zoneKey": "TR",
-            "currency": "TRY",
-            "datetime": arrow.get(datetime.combine(date, time))
-            .to("Europe/Istanbul")
-            .datetime,
-            "price": price,
+) -> list:
+    if target_datetime is None:
+        target_datetime = datetime.now(tz=TR_TZ) - timedelta(hours=2)
```

Comment:
Is this because there is a delay?

Reply:
actually this was to try to remove the last point of data which was rubbish but it's not necessary since there is data validation in fetch_production so i'll remove